lib/auth/auth_provider.js

```diff
@@ -0,0 +1,177 @@
+'use strict';
+
+// const MongoCredentials = require('./mongo_credentials').MongoCredentials;
+const MongoError = require('../error').MongoError;
+const Query = require('../connection/commands').Query;
+
+/**
+ * Creates a new Authentication mechanism
+ * @class
+ */
+class AuthProvider {
+  constructor(bson) {
+    this.bson = bson;
+    this.authStore = [];
+  }
+
+  _buildCommand(namespace, cmd) {
+    return new Query(this.bson, namespace, cmd, {
+      numberToSkip: 0,
+      numberToReturn: 1
+    });
+  }
+
+  /**
+   * Authenticate
+   * @method
+   * @param {AuthWriteCommand} writeCommand Topology the authentication method is being called on
+   * @param {Connection[]} connections Connections to authenticate using this authenticator
+   * @param {MongoCredentials} credentials Authentication credentials
+   * @param {authResultCallback} callback The callback to return the result from the authentication
+   */
+  auth(writeCommand, connections, credentials, callback) {
```

Comment:
At this point in time, `authProviders` are shared across an entire topology. So every server (and by extension every pool) inherits an authProvider from the parent replset/mongos. With the way it is currently done, I don't think we can safely do what you are suggesting, but it may be possible in follow-on work.